src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,64 +4,166 @@
     {
         if (!isCompatibleDevicewise(layouts[i]))
-            return false;
+        {
+            if (logger)
+                logger->log("Device-Incompatible descriptor set layout found at index %u. Sets for the layouts following this index will not be created.", system::ILogger::ELL_WARNING, i);
+
+            break;
+        }
         
-        descriptorOffsets[i] = allocateDescriptorOffsets(layouts[i]);
+        if (!allocateDescriptorOffsets(descriptorOffsets.emplace_back(), layouts[i]))
+        {
+            if (logger)
+                logger->log("Failed to allocate descriptor offsets in the pool's storage for descriptor set layout at index %u. Sets for the layouts following this index will not be created.", system::ILogger::ELL_WARNING, i);
+
+            descriptorOffsets.pop_back();
+            break;
+        }
+    }
+
+    const auto successCount = descriptorOffsets.size();
+    assert(count >= successCount);
+    std::fill_n(output + successCount, count - successCount, nullptr);
+
+    const uint32_t allocatedDSOffset = m_descriptorSetAllocator.alloc_addr(successCount, 1u);
+    if (allocatedDSOffset != m_descriptorSetAllocator.invalid_address)
+    {
+        if (createDescriptorSets_impl(successCount, layouts, descriptorOffsets.data(), allocatedDSOffset, output))
+        {
+            for (uint32_t i = 0u; i < successCount; ++i)
+                m_allocatedDescriptorSets[allocatedDSOffset + i] = output[i].get();
+
+            return successCount;
+        }
+    }
+    else
+    {
+        if (logger)
+            logger->log("Failed to allocate descriptor sets.", system::ILogger::ELL_ERROR);
     }
 
-    return createDescriptorSets_impl(count, layouts, descriptorOffsets.data(), output);
+    // Free the allocated offsets for all the successfully allocated descriptor sets.
+    for (uint32_t i = 0u; i < successCount; ++i)
+        freeDescriptorOffsets(descriptorOffsets[i], layouts[i]);
+
+    return 0;
 }
 
 bool IDescriptorPool::reset()
 {
-    const bool allowsFreeing = m_creationParameters.flags.hasFlags(ECF_FREE_DESCRIPTOR_SET_BIT);
-    for (uint32_t t = 0; t < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++t)
+    for (const uint32_t setIndex : m_descriptorSetAllocator)
     {
-        if (!m_descriptorAllocators[t])
-            continue;
+        deleteSetStorage(m_allocatedDescriptorSets[setIndex]);
 
-        // TODO(achal): I should call the dtors on descriptors only after calling vkResetDescriptorPool (in reset_impl)
-        const uint32_t allocatedCount = m_descriptorAllocators[t]->getAllocatedDescriptorCount(allowsFreeing);
-        std::destroy_n(getDescriptorStorage(static_cast<asset::IDescriptor::E_TYPE>(t)), allocatedCount);
-
-        m_descriptorAllocators[t]->reset(allowsFreeing);
+        m_allocatedDescriptorSets[setIndex]->m_pool = nullptr;
+        m_allocatedDescriptorSets[setIndex]->m_poolOffset = ~0u;
+        m_allocatedDescriptorSets[setIndex] = nullptr;
```

Comment:
this should be folded into `deleteSetStorage`

Reply:
Oh okay.
Previously we discussed to keep it outside (in this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1118753183), so I did that.

```C++
    // basically what `~IGPUDescriptorSet` calls minus the free-ing of the set itself
    deleteSetStorage(m_allocatedDescriptorSets[setIndex]);
```

Reply:
Resolved.

Reply:
> Previously we discussed to keep it outside (in this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1118753183), so I did that.

"minus the free-ing of the set itself" - I meant `vkFreeDescriptorSets` by that, sorry I wasn't clear